cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -1304,27 +1304,43 @@ class ObjectOutputStream final : public io::OutputStream {
       return Status::Invalid("Operation on closed stream");
     }
 
-    if (!current_part_ && nbytes >= part_upload_threshold_) {
-      // No current part and data large enough, upload it directly
-      // (without copying if the buffer is owned)
-      RETURN_NOT_OK(UploadPart(data, nbytes, owned_buffer));
-      pos_ += nbytes;
-      return Status::OK();
+    const int8_t* data_ptr = reinterpret_cast<const int8_t*>(data);
+    int64_t offset = 0;
```

Comment:
Instead of maintaining a separate offset, you can simply bump `data_ptr` and decrease `nbytes` as needed.